<a href="https://colab.research.google.com/github/ansiyo/ML-For-Beginners/blob/main/Demo_53AModel_Evaluation_Metrics.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h2> Demo: Model Evaluation Metrics for Regression </h2>

In this demo, you will learn how to use different model evaluation metrics like MAPE, RMSE, RMSLE for regression model.

# Problem Statement

The dataset has various features from coffee beans from different countries. The aim is to predict mean altitude.

# Importing Libraries

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split,GridSearchCV
from sklearn.metrics import mean_squared_error,mean_squared_log_error
from sklearn.linear_model import LinearRegression
import matplotlib.pyplot as plt

# Data Analysis

In [2]:
data=pd.read_csv('/content/arabica_data.csv')

In [3]:
data.head()

Country.of.Origin  Number.of.Bags  Bag.Weight  Aroma  Flavor  Aftertaste  \
0          Ethiopia             300          60   8.67    8.83        8.67   
1          Ethiopia             300          60   8.75    8.67        8.50   
2         Guatemala               5           1   8.42    8.50        8.42   
3          Ethiopia             320          60   8.17    8.58        8.42   
4          Ethiopia             300          60   8.25    8.50        8.25   

   Acidity  Body  Balance  Uniformity  Clean.Cup  Sweetness  Cupper.Points  \
0     8.75  8.50     8.42        10.0       10.0       10.0           8.75   
1     8.58  8.42     8.42        10.0       10.0       10.0           8.58   
2     8.42  8.33     8.42        10.0       10.0       10.0           9.25   
3     8.42  8.50     8.25        10.0       10.0       10.0           8.67   
4     8.50  8.42     8.33        10.0       10.0       10.0           8.58   

   Total.Cup.Points  Moisture  altitude_mean_meters  
0             90.58      0.12                2075.0  
1             89.92      0.12                2075.0  
2             89.75      0.00                1700.0  
3             89.00      0.11                2000.0  
4             88.83      0.12                2075.0

In [4]:
data.shape

(1311, 16)

<div style="background-color:##FBFFE3">
    <div style="float:left">
        <img src='https://drive.google.com/uc?export=download&id=1gpc8kbwaenTpaIvnO_MQR3vw12rKYQrJ', width=20,height=20,>
    </div>
    <span style="padding:20px;">Dataset has 1310 rows and 16 columns</span>
</div>

<div style="background-color:##EDF9FF">
  <ul><li><h3> Testing For Null Values</h3></li></ul>
</div>

In [5]:
pd.DataFrame({'Count':data.isnull().sum(),'Percentage':data.isnull().sum()/data.shape[0]})

Count  Percentage
Country.of.Origin         1    0.000763
Number.of.Bags            0    0.000000
Bag.Weight                0    0.000000
Aroma                     0    0.000000
Flavor                    0    0.000000
Aftertaste                0    0.000000
Acidity                   0    0.000000
Body                      0    0.000000
Balance                   0    0.000000
Uniformity                0    0.000000
Clean.Cup                 0    0.000000
Sweetness                 0    0.000000
Cupper.Points             0    0.000000
Total.Cup.Points          0    0.000000
Moisture                  0    0.000000
altitude_mean_meters    227    0.173150

<div style="background-color:##FBFFE3">
    <div style="float:left">
        <img src='https://drive.google.com/uc?export=download&id=1gpc8kbwaenTpaIvnO_MQR3vw12rKYQrJ', width=20,height=20,>
    </div>
    <span style="padding:20px;">Dataset has null values. Since <b>Country.of.Origin</b> has only one null values we can drop that row and <b>altitube_mean_meters</b> has 227 null values we will have to impute them.</span>
</div>

<div style="background-color:##EDF9FF">
  <ul><li><h3>Handling Null Values</h3></li></ul>
</div>

In [6]:
# Null value in Country of origin
data[data['Country.of.Origin'].isna()==True]

Country.of.Origin  Number.of.Bags  Bag.Weight  Aroma  Flavor  Aftertaste  \
1197               NaN             149          70   6.75    6.75        6.42   

      Acidity  Body  Balance  Uniformity  Clean.Cup  Sweetness  Cupper.Points  \
1197     6.83  7.58      7.5        10.0       10.0       10.0           7.25   

      Total.Cup.Points  Moisture  altitude_mean_meters  
1197             79.08       0.1                   NaN

In [7]:
data=data.drop(1197,axis=0)

In [8]:
#Handling target null value
mean=data.altitude_mean_meters.mean()
data.altitude_mean_meters=data.altitude_mean_meters.fillna(mean)

In [9]:
pd.DataFrame({'Count':data.isnull().sum(),'Percentage':data.isnull().sum()/data.shape[0]})

Count  Percentage
Country.of.Origin         0         0.0
Number.of.Bags            0         0.0
Bag.Weight                0         0.0
Aroma                     0         0.0
Flavor                    0         0.0
Aftertaste                0         0.0
Acidity                   0         0.0
Body                      0         0.0
Balance                   0         0.0
Uniformity                0         0.0
Clean.Cup                 0         0.0
Sweetness                 0         0.0
Cupper.Points             0         0.0
Total.Cup.Points          0         0.0
Moisture                  0         0.0
altitude_mean_meters      0         0.0

<div style="background-color:##FBFFE3">
    <div style="float:left">
        <img src='https://drive.google.com/uc?export=download&id=1gpc8kbwaenTpaIvnO_MQR3vw12rKYQrJ', width=20,height=20,>
    </div>
    <span style="padding:20px;">Dataset has no null values now.</span>
</div>

In [10]:
print(*data.columns,sep=',')

Country.of.Origin,Number.of.Bags,Bag.Weight,Aroma,Flavor,Aftertaste,Acidity,Body,Balance,Uniformity,Clean.Cup,Sweetness,Cupper.Points,Total.Cup.Points,Moisture,altitude_mean_meters


<div style="background-color:##EDF9FF">
  <ul><li><h3>Selecting Features</h3></li></ul>
</div>

In [11]:
X=data[['Aroma','Flavor','Aftertaste','Acidity','Body','Balance','Uniformity','Clean.Cup',\
             'Sweetness','Cupper.Points','Total.Cup.Points','Moisture']]
y=data.altitude_mean_meters

<div style="background-color:##EDF9FF">
  <ul><li><h3>Training And Testing Data</h3></li></ul>
</div>

In [12]:
X_train, X_test, y_train, y_test=train_test_split(X,y,random_state=2,test_size=0.3)

# Training Model

In [13]:
lreg=LinearRegression()
lreg.fit(X_train,y_train)
lpred=lreg.predict(X_test)


In [14]:
#MAPE
mape= np.mean(np.abs((np.array(y_test) - np.array(lpred)) / y_test)) * 100
print(mape)

1839.5565935584336


In [15]:
# MSE
mean_squared_error(y_test,lpred)

1679073.41987456

In [16]:
# RMSE
np.sqrt(mean_squared_error(y_test,lpred))

1295.7906543398744

In [17]:
# MSLE
mean_squared_log_error(y_test,lpred)

0.975258539720536

In [18]:
#RMLSE
np.sqrt(mean_squared_log_error(y_test,lpred))

0.987551790905437

*__Conclusion__*: In this demonstration, we have learned how to use regression metrics.

In [19]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split,GridSearchCV
from sklearn.metrics import mean_squared_error,mean_squared_log_error
from sklearn.linear_model import LinearRegression
import matplotlib.pyplot as plt

In [30]:
data=pd.read_csv('/content/wine.csv')

In [31]:
data.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            7.2              0.23         0.32             8.5      0.058   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 47.0                 186.0   0.9956  3.19       0.40   

   alcohol  quality  
0      8.8        6  
1      9.5        6  
2     10.1        6  
3      9.9        6  
4      9.9        6

In [32]:
data.shape

(4898, 12)

In [33]:
pd.DataFrame({'Count':data.isnull().sum(),'Percentage':data.isnull().sum()/data.shape[0]})

Count  Percentage
fixed acidity             0         0.0
volatile acidity          0         0.0
citric acid               0         0.0
residual sugar            0         0.0
chlorides                 0         0.0
free sulfur dioxide       0         0.0
total sulfur dioxide      0         0.0
density                   0         0.0
pH                        0         0.0
sulphates                 0         0.0
alcohol                   0         0.0
quality                   0         0.0

In [35]:
result = data.dtypes
print(result)

fixed acidity           float64
volatile acidity        float64
citric acid             float64
residual sugar          float64
chlorides               float64
free sulfur dioxide     float64
total sulfur dioxide    float64
density                 float64
pH                      float64
sulphates               float64
alcohol                 float64
quality                   int64
dtype: object


In [36]:
print(*data.columns,sep=',')


fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality


In [ ]:
X=data[['fixed acidity','volatile acidity','citric acid','residual sugar','chlorides','sulfur dioxide','total sulfur dioxide','density',\
             'pH','sulphates','alcohol','quality']]
y=data.alcohol

In [46]:
X_train, X_test, y_train, y_test=train_test_split(X,y,random_state=2,test_size=0.3)

In [47]:
lreg=LinearRegression()
lreg.fit(X_train,y_train)
lpred=lreg.predict(X_test)

In [48]:
mape= np.mean(np.abs((np.array(y_test) - np.array(lpred)) / y_test)) * 100
print(mape)

1839.5565935584336


In [49]:
mean_squared_error(y_test,lpred)

1679073.41987456

In [50]:
np.sqrt(mean_squared_error(y_test,lpred))

1295.7906543398744

In [51]:
mean_squared_log_error(y_test,lpred)

0.975258539720536

In [52]:
np.sqrt(mean_squared_log_error(y_test,lpred))

0.987551790905437

In [34]:
# return the dtype of each column
result = df.dtypes

# Print the result
print(result)

NameError: name 'df' is not defined